// testcafe/lib/utils/handle-errors.js
'use strict';

exports.__esModule = true;

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

exports.registerErrorHandlers = registerErrorHandlers;
exports.addRunningTest = addRunningTest;
exports.removeRunningTest = removeRunningTest;
exports.startHandlingTestErrors = startHandlingTestErrors;
exports.stopHandlingTestErrors = stopHandlingTestErrors;

var _testRun = require('../errors/test-run');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const runningTests = {};
let handlingTestErrors = false;

function handleError(ErrorCtor, message) {
    if (handlingTestErrors) {
        (0, _values2.default)(runningTests).forEach(testRun => {
            testRun.addError(new ErrorCtor(message));

            removeRunningTest(testRun);
        });
    } else {
        /* eslint-disable no-process-exit */
        /* eslint-disable no-console */
        console.log(message);

        setTimeout(() => process.exit(1), 0);
        /* eslint-enable no-process-exit */
        /* eslint-enable no-console */
    }
}

function formatUnhandledRejectionReason(reason) {
    const reasonType = typeof reason;
    const isPrimitiveType = reasonType !== 'object' && reasonType !== 'function';

    if (isPrimitiveType) return String(reason);

    if (reason instanceof Error) return reason.stack;

    return Object.prototype.toString.call(reason);
}

function onUnhandledRejection(reason) {
    if (reason && reason.isRejectedDriverTask) return;

    const message = formatUnhandledRejectionReason(reason);

    handleError(_testRun.UnhandledPromiseRejectionError, message);
}

function onUncaughtException(err) {
    handleError(_testRun.UncaughtExceptionError, err.stack);
}

function registerErrorHandlers() {
    process.on('unhandledRejection', onUnhandledRejection);
    process.on('uncaughtException', onUncaughtException);
}

function addRunningTest(testRun) {
    runningTests[testRun.id] = testRun;
}

function removeRunningTest(testRun) {
    delete runningTests[testRun.id];
}

function startHandlingTestErrors() {
    handlingTestErrors = true;
}

function stopHandlingTestErrors() {
    handlingTestErrors = false;
}
