// testcafe/lib/utils/moment-loader.js
'use strict';

exports.__esModule = true;

var _resolveFrom = require('resolve-from');

var _resolveFrom2 = _interopRequireDefault(_resolveFrom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MOMENT_MODULE_NAME = 'moment';
const DURATION_FORMAT_MODULE_NAME = 'moment-duration-format';

function restoreInitialCacheState(module, path) {
    if (module) require.cache[path] = module;else delete require.cache[path];
}

function getSideMomentModulePath(sidePath) {
    try {
        return (0, _resolveFrom2.default)(sidePath, MOMENT_MODULE_NAME);
    } catch (err) {
        return '';
    }
}

function getModulesPaths() {
    const durationFormatModulePath = require.resolve(DURATION_FORMAT_MODULE_NAME);

    return {
        durationFormatModulePath,

        mainMomentModulePath: require.resolve(MOMENT_MODULE_NAME),
        sideMomentModulePath: getSideMomentModulePath(durationFormatModulePath)
    };
}

function getCachedAndCleanModules(modulePath) {
    const cachedModule = require.cache[modulePath];

    delete require.cache[modulePath];

    require(modulePath);

    return { cachedModule, cleanModule: require.cache[modulePath] };
}

function getMomentModules({ mainMomentModulePath, sideMomentModulePath }) {
    var _getCachedAndCleanMod = getCachedAndCleanModules(mainMomentModulePath);

    const cachedModule = _getCachedAndCleanMod.cachedModule,
          cleanModule = _getCachedAndCleanMod.cleanModule;


    return {
        sideModule: require.cache[sideMomentModulePath],
        mainModule: cleanModule,
        cachedModule
    };
}

function getMomentModuleWithDurationFormatPatch() {
    const modulesPaths = getModulesPaths();
    const momentModules = getMomentModules(modulesPaths);

    const sideMomentModulePath = modulesPaths.sideMomentModulePath,
          mainMomentModulePath = modulesPaths.mainMomentModulePath,
          durationFormatModulePath = modulesPaths.durationFormatModulePath;


    if (sideMomentModulePath && sideMomentModulePath !== mainMomentModulePath) {
        require.cache[sideMomentModulePath] = momentModules.mainModule;

        require(durationFormatModulePath);

        restoreInitialCacheState(momentModules.sideModule, sideMomentModulePath);
    } else {
        const durationFormatSetup = require(durationFormatModulePath);

        if (!sideMomentModulePath) durationFormatSetup(momentModules.mainModule.exports);
    }

    restoreInitialCacheState(momentModules.cachedModule, mainMomentModulePath);

    return momentModules.mainModule.exports;
}

exports.default = getMomentModuleWithDurationFormatPatch();
module.exports = exports['default'];
